### ISSUE TITLE
[Help wanted] Application crashed unexpectedly

### ISSUE THREAD

**Issue by Dragon1573** [opentoonz/opentoonz#5171]
Manual Validation ended with: 
![image](https://github.com/microsoft/winget-pkgs/assets/114952565/12e1bd45-39ad-466c-86fc-9c03c9a610e6)

Edit: adding Crash file and Mini Dump file.
- [Crash-20231114-104145.dmp](https://github.com/microsoft/winget-pkgs/files/13355842/Crash-20231114-104145.dmp)
- [Crash-20231114-104145.log](https://github.com/microsoft/winget-pkgs/files/13355843/Crash-20231114-104145.log)

```
OpenToonz 1.7.1  (Build Mar 14 
> 2023)

Report Date: 
> 2023-11-14 10:33:05
Crash Reason: EXCEPTION_ACCESS_VIOLATION

Build ABI: x86_64-little_endian-llp64
Operating System: Windows 10 Version 2009
OS Kernel: 10.0.19044
CPU Threads: 8


Crash File: C:\Users\User\AppData\Local\OpenToonz\OpenToonz\crash\Crash-
> 20231114-103305.log
Mini Dump File: C:\Users\User\AppData\Local\OpenToonz\OpenToonz\crash\Crash-
> 20231114-103305.dmp

==== Modules ====
C:\Program Files\OpenToonz\OpenToonz.exe
C:\Windows\SYSTEM32\ntdll.dll
C:\Windows\System32\KERNEL32.DLL
C:\Windows\System32\KERNELBASE.dll
C:\Windows\System32\USER32.dll
C:\Windows\System32\win32u.dll
C:\Program Files\OpenToonz\Qt5Xml.dll
C:\Program Files\OpenToonz\Qt5PrintSupport.dll
C:\Windows\System32\GDI32.dll
C:\Windows\SYSTEM32\OPENGL32.dll
C:\Windows\System32\gdi32full.dll
C:\Windows\System32\msvcp_win.dll
C:\Windows\System32\ucrtbase.dll
C:\Windows\System32\msvcrt.dll
C:\Windows\System32\SHELL32.dll
C:\Windows\System32\ADVAPI32.dll
C:\Windows\System32\COMDLG32.dll
C:\Program Files\OpenToonz\glut64.dll
C:\Windows\System32\combase.dll
C:\Windows\System32\ole32.dll
C:\Windows\System32\RPCRT4.dll
C:\Program Files\OpenToonz\Qt5SerialPort.dll
C:\Windows\System32\OLEAUT32.dll
C:\Windows\System32\shcore.dll
C:\Windows\System32\sechost.dll
C:\Windows\System32\SETUPAPI.dll
C:\Windows\System32\SHLWAPI.dll
C:\Program Files\OpenToonz\EDSDK.dll
C:\Windows\System32\cfgmgr32.dll
C:\Windows\System32\bcrypt.dll
C:\Windows\System32\WS2_32.dll
C:\Program Files\OpenToonz\colorfx.dll
C:\Program Files\OpenToonz\image.dll
C:\Program Files\OpenToonz\tnztools.dll
C:\Program Files\OpenToonz\tnzstdfx.dll
C:\Program Files\OpenToonz\Qt5Widgets.dll
C:\Program Files\OpenToonz\tfarm.dll
C:\Program Files\OpenToonz\Qt5Gui.dll
C:\Program Files\OpenToonz\Qt5Core.dll
C:\Program Files\OpenToonz\toonzqt.dll
C:\Program Files\OpenToonz\sound.dll
C:\Program Files\OpenToonz\opencv_world451.dll
C:\Program Files\OpenToonz\toonzlib.dll
C:\Program Files\OpenToonz\Qt5Script.dll
C:\Program Files\OpenToonz\tnzext.dll
C:\Program Files\OpenToonz\tnzbase.dll
C:\Program Files\OpenToonz\tnzcore.dll
C:\Program Files\OpenToonz\Qt5OpenGL.dll
C:\Program Files\OpenToonz\Qt5Multimedia.dll
C:\Program Files\OpenToonz\Qt5Network.dll
C:\Windows\SYSTEM32\GLU32.dll
C:\Windows\SYSTEM32\MPR.dll
C:\Windows\System32\CRYPT32.dll
C:\Windows\SYSTEM32\dbghelp.dll
C:\Program Files\OpenToonz\MSVCP140.dll
C:\Program Files\OpenToonz\MSVCP140_1.dll
C:\Program Files\OpenToonz\VCRUNTIME140.dll
C:\Windows\SYSTEM32\WINSPOOL.DRV
C:\Windows\SYSTEM32\WINMM.dll
C:\Program Files\OpenToonz\VCRUNTIME140_1.dll
C:\Windows\SYSTEM32\MFPlat.DLL
C:\Windows\SYSTEM32\MF.dll
C:\Windows\WinSxS\amd64_microsoft.windows.common-controls_6595b64144ccf1df_5.82.19041.1110_none_792d1c772443f647\COMCTL32.dll
C:\Windows\SYSTEM32\MSVFW32.dll
C:\Windows\SYSTEM32\VERSION.dll
C:\Windows\SYSTEM32\AVIFIL32.dll
C:\Program Files\OpenToonz\libmypaint-1-4-0.dll
C:\Program Files\OpenToonz\glew32.dll
C:\Windows\SYSTEM32\d3d11.dll
C:\Windows\SYSTEM32\dxgi.dll
C:\Windows\SYSTEM32\USERENV.dll
C:\Windows\SYSTEM32\NETAPI32.dll
C:\Program Files\OpenToonz\Qt5Svg.dll
C:\Windows\SYSTEM32\dwmapi.dll
C:\Windows\SYSTEM32\UxTheme.dll
C:\Program Files\OpenToonz\CONCRT140.dll
C:\Windows\SYSTEM32\IPHLPAPI.DLL
C:\Windows\SYSTEM32\DNSAPI.dll
C:\Windows\SYSTEM32\MSACM32.dll
C:\Program Files\OpenToonz\libjson-c-2.dll
C:\Program Files\OpenToonz\libintl-8.dll
C:\Program Files\OpenToonz\libiconv-2.dll
C:\Windows\SYSTEM32\winmmbase.dll
C:\Windows\SYSTEM32\NETUTILS.DLL
C:\Windows\SYSTEM32\dbgcore.DLL
C:\Windows\SYSTEM32\SRVCLI.DLL
C:\Windows\SYSTEM32\MFCORE.DLL
C:\Windows\SYSTEM32\powrprof.dll
C:\Windows\SYSTEM32\ksuser.dll
C:\Windows\SYSTEM32\kernel.appcore.dll
C:\Windows\SYSTEM32\CRYPTBASE.DLL
C:\Windows\System32\IMM32.DLL
C:\Windows\System32\NSI.dll
C:\Windows\System32\bcryptPrimitives.dll
C:\Windows\SYSTEM32\RTWorkQ.DLL
C:\Windows\SYSTEM32\UMPDC.dll
C:\Program Files\OpenToonz\platforms\qwindows.dll
C:\Windows\SYSTEM32\WTSAPI32.dll
C:\Windows\SYSTEM32\windows.storage.dll
C:\Windows\SYSTEM32\Wldp.dll
C:\Windows\SYSTEM32\profapi.dll
C:\Program Files\OpenToonz\styles\qwindowsvistastyle.dll
C:\Program Files\OpenToonz\iconengines\qsvgicon.dll
C:\Windows\system32\d3d9.dll
C:\Windows\System32\MSCTF.dll
C:\Windows\SYSTEM32\textinputframework.dll
C:\Windows\System32\CoreUIComponents.dll
C:\Windows\SYSTEM32\ntmarta.dll
C:\Windows\System32\CoreMessaging.dll
C:\Windows\SYSTEM32\wintypes.dll
C:\Windows\System32\clbcatq.dll
C:\Windows\System32\AppXDeploymentClient.dll
C:\Windows\system32\dataexchange.dll
C:\Windows\system32\dcomp.dll
C:\Windows\system32\twinapi.appcore.dll
C:\Windows\system32\dwrite.dll
==== End ====

==== Backtrace ====
  1> RtlDeleteAce [0x7ffa01388a3c] <ntdll.dll>
  2> RtlRaiseException [0x7ffa01361276] <ntdll.dll>
  3> KiUserExceptionDispatcher [0x7ffa013b0bfe] <ntdll.dll>
  4> TColorStyle::getRasterStyleFx [0x7ff9c22f5f49] <tnzcore.dll>
  5> TOStream::openChild [0x7ff9c249f050] <tnzcore.dll>
  6> TProject::save [0x7ff9c35947f9] <toonzlib.dll>
  7> TProjectManager::createSandboxIfNeeded [0x7ff9c358e0b0] <toonzlib.dll>
  8> TSmartPointerT<TXshSoundLevel>::~TSmartPointerT<TXshSoundLevel> [0x7ff68a0e0e53] <OpenToonz.exe>
  9> TSmartPointerT<TXshSoundLevel>::~TSmartPointerT<TXshSoundLevel> [0x7ff68a0e2d41] <OpenToonz.exe>
 10> TSmartPointerT<TXshSoundTextLevel>::~TSmartPointerT<TXshSoundTextLevel> [0x7ff68a50eab7] <OpenToonz.exe>
 11> TSmartPointerT<TXshSoundTextLevel>::~TSmartPointerT<TXshSoundTextLevel> [0x7ff68a50e092] <OpenToonz.exe>
 12> BaseThreadInitThunk [0x7ffa00197034] <KERNEL32.DLL>
 13> RtlUserThreadStart [0x7ffa01362651] <ntdll.dll>
==== End ====


```

_Originally posted by stephengillie at https://github.com/microsoft/winget-pkgs/pull/125044#issuecomment-1810908389_

-----

Hello contributors,

I'm making efforts on creating and submitting manifests of this package for [microsoft/winget-pkgs](https://github.com/microsoft/winget-pkgs). One of the moderators manually validated this package and meet a crash.

Crash logs are attached above.

**Comment by RodneyBaker** [comment]
Try uninstalling and then installing the latest nightly release located at:

https://github.com/opentoonz/opentoonz_nightlies/releases

**Comment by Dragon1573** [comment]
After his/her manual validation, nightly build also have the same problem.

-----

Unfortunately, the Nightly has the same error as the [other version](https://github.com/microsoft/winget-pkgs/pull/125044):

![image](https://github.com/microsoft/winget-pkgs/assets/114952565/d79c20ce-2254-4d57-80d2-6c8f7cc3cd9d)


_Originally posted by @stephengillie in https://github.com/microsoft/winget-pkgs/pull/126821#issuecomment-1814981766_

**Comment by RodneyBaker** [comment]
Ah, actually a different error.
You are missing the glut32.dll (driver).
Copy that into your opentooz directory (or other location as appropriate and all should be well.

[glut32.zip](https://github.com/opentoonz/opentoonz/files/13390380/glut32.zip)

If you can't find a copy I've added it here in a zip file which you will need to unzip into your Opentoonz program directory:  C:\Program Files\Opentoonz

**Comment by Dragon1573** [comment]
It seems `glut32.dll` is provided by another package? Is there any installer wizard for that? We can try if adding it as a dependency can solve the problem.

**Comment by stephengillie** [comment]
This file `glut32.dll` appears to be in the [OpenGL Utility Toolkit](https://www.opengl.org/resources/libraries/glut/glut_downloads.php). I'm not seeing it in WinGet's public manifests currently, so it could be added in the same way as this application. Once added it can be referenced as a dependency in this application's manifest. 

These screenshots are from my [scripted](https://github.com/microsoft/winget-pkgs/blob/master/Tools/ManualValidationPipeline.ps1) validation system. 
- This system installs, runs every EXE, then performs a Defender scan. 
- The `t32bitsrv.exe` error occurred when every EXE was launched, but not after. 
  - I originally didn't include this error as it didn't occur when double-clicking the desktop icon. 
  - Frequently, various developers seem to include EXEs that aren't meant to be executed directly, for various reasons, so it seemed like another case of this. 
  - Is this EXE supposed to be running in the background? I believe Application Log errors mentioned a missing server. But again, this is sometimes a result of launching EXEs that weren't intended for direct launching.

**Comment by Dragon1573** [comment]
> I'm not seeing it in WinGet's public manifests currently, so it could be added in the same way as this application.

I think it can't ...

OpenGL Utility Toolkit only provides ZIP archives for Windows, and there's no install wizard file of this package. `*.h`, `*.lib` and `*.dll` files from `glutdlls37beta.zip` should be placed into different folder seperately.

I previously meet a similar package like this. Package A requires a DLL file, which can be provided by package B. Package B can be easily installed with an MSI wizard. But in this issue case, package B can't be easily installed as package A and even unable to create a manifest for it ...

**Comment by Dragon1573** [comment]
### :loud_speaker: Notes

The nightly build of this package is accepted and successfully merged by #126821.

As we are unable to add OpenGL Utilities Toolkit to package repository, `Opentoonz.Opentoonz` will always pop up a warning when running in validation pipeline.

This behaviour is expected. The issue can be closed.
